tempesta_fw/apm.c

```diff
@@ -1058,6 +1215,149 @@ tfw_apm_del_srv(TfwServer *srv)
 	srv->apmref = NULL;
 }
 
+void
+tfw_apm_hm_srv_rcount_update(TfwStr *uri_path, void *apmref)
+{
+	TfwApmHMCtl *hmctl = &((TfwApmData *)apmref)->hmctl;
+	tfw_str_eq_flags_t flags = TFW_STR_EQ_CASEI;
+
+	BUG_ON(!hmctl->hm);
+	if (tfw_str_eq_cstr(uri_path, hmctl->hm->url, hmctl->hm->urlsz, flags))
+		atomic64_inc(&hmctl->rcount);
+}
+
+bool
+tfw_apm_hm_srv_alive(int status, TfwStr *body, void *apmref)
+{
+	u32 crc32;
+	TfwApmHM *hm = ((TfwApmData *)apmref)->hmctl.hm;
+
+	BUG_ON(!hm);
+	if (hm->codes && test_bit(HTTP_CODE_BIT_NUM(status), hm->codes))
+		return true;
+
+	if (hm->crc32
+	    && body->len
+	    && !tfw_str_crc32_calc(body, &crc32)
+	    && crc32 == hm->crc32)
+		return true;
+
+	return false;
+}
+
+bool
+tfw_apm_hm_srv_limit(int status, void *apmref)
+{
+	unsigned int i, sum = 0;
+	TfwApmHMStats *hmstats = ((TfwApmData *)apmref)->hmctl.hmstats;
+	TfwApmHMHistory *history = NULL;
+	TfwApmHMCfg *cfg = NULL;
+	unsigned long ts;
+
+	BUG_ON(!hmstats);
+	for (i = 0; i < tfw_hm_codes_cnt; ++i) {
+		if (hmstats[i].hmcfg->code == status ||
+		    hmstats[i].hmcfg->code == (int)(status / 100))
+		{
+			history = hmstats[i].history;
+			cfg = hmstats[i].hmcfg;
+			hmstats = &hmstats[i];
+			break;
+		}
+	}
+
+	if (!history)
+		return false;
+
+	BUG_ON(!cfg);
+	spin_lock(&hmstats->lock);
+
+	ts = jiffies * HM_FREQ / (cfg->tframe * HZ);
+	i = ts % HM_FREQ;
+	if (history[i].ts != ts) {
+		history[i].ts = ts;
+		history[i].resp = 0;
+	}
+	++history[i].resp;
+	for (i = 0; i < HM_FREQ; ++i)
+		if (history[i].ts + HM_FREQ > ts)
+			sum += history[i].resp;
+
+	WRITE_ONCE(hmstats->rsum, sum);
+	spin_unlock(&hmstats->lock);
+
+	if (sum > cfg->limit)
+		return true;
+
+	return false;
+}
+
+bool
+tfw_apm_hm_set_srv(const char *name, TfwServer *srv)
+{
+	TfwApmHMCtl *hmctl;
+	TfwApmHM *hm;
+	unsigned long now;
+
+	BUG_ON(!srv->apmref);
+	hmctl = &((TfwApmData *)srv->apmref)->hmctl;
+	list_for_each_entry(hm, &tfw_hm_list, list) {
+		if(!strcasecmp(name, hm->name)) {
```

Comment:
Please place a space between `if` and opening brace. Also it's better to write such loops like
```C
if (strcasecmp(name, hm->name))
     continue;
// the if body
```

Reply:
Corrected.